src/aspire/abinitio/commonline_c3_c4.py

```diff
@@ -231,11 +249,153 @@ def _estimate_third_rows(self, vijs, viis):
+        # the corresponding third rows vis.
+        Ri_tildes = np.array([self.complete_third_row_to_rot(vi) for vi in vis])
+
+        # Step 2: Construct all in-plane rotation matrices, R_theta_ijs.
+        max_angle = (360 // order) * order
+        theta_ijs = np.arange(0, max_angle, degree_res) * np.pi / 180
+        R_theta_ijs = Rotation.about_axis("z", theta_ijs, dtype=self.dtype).matrices
+
+        # Step 3: Compute the correlation over all shifts.
+        # Generate shifts.
+        r_max = pf.shape[0]
+        shifts, shift_phases, _ = self._generate_shift_phase_and_filter(
+            r_max, max_shift_1d, shift_step
+        )
+        n_shifts = len(shifts)
+        all_shift_phases = shift_phases.T
+
+        n_pairs = n_img * (n_img - 1) // 2
+        max_corrs = np.zeros(n_pairs)
+        max_corrs_idx = np.zeros(n_pairs)
+
+        # Q is the n_img x n_img  Hermitian matrix defined by Q = q*q^H,
+        # where q = (exp(i*order*theta_0), ..., exp(i*order*theta_{n_img-1}))^H,
+        # and theta_i in [0, 2pi/order) is the in-plane rotation angle for the i'th image.
+        Q = np.zeros((n_img, n_img), dtype=complex)
+
+        # Transpose pf and reconstruct the full polar Fourier for use in correlation.
+        # self.pf only consists of rays in the range [180, 360).
+        pf = pf.transpose((2, 1, 0))
+        pf = np.concatenate((pf, np.conj(pf)), axis=1)
+
+        # Normalize rays.
+        pf /= norm(pf, axis=-1)[..., np.newaxis]
+
+        with tqdm(total=n_pairs) as pbar:
+            idx = 0
+            # Note: the ordering of i and j in these loops should not be changed as
+            # they correspond to the ordered tuples (i, j), for i<j.
+            for i in range(n_img):
+                pf_i = pf[i]
+
+                # Generate shifted versions of images.
+                pf_i_shifted = np.array(
+                    [pf_i * shift_phase for shift_phase in all_shift_phases]
+                )
+
+                Ri_tilde = Ri_tildes[i]
+
+                for j in range(i + 1, n_img):
+                    pf_j = pf[j]
+
+                    Rj_tilde = Ri_tildes[j]
+
+                    # Compute all possible rotations between the i'th and j'th images.
+                    Us = np.array(
+                        [
+                            Ri_tilde.T @ R_theta_ij @ Rj_tilde
+                            for R_theta_ij in R_theta_ijs
+                        ]
+                    )
+
+                    # Find the angle between common lines induces by the rotations.
+                    c1s = np.array([[-U[1, 2], U[0, 2]] for U in Us])
+                    c2s = np.array([[U[2, 1], -U[2, 0]] for U in Us])
+
+                    # Convert from angles to indices.
+                    c1s = self.clAngles2Ind(c1s, n_theta)
+                    c2s = self.clAngles2Ind(c2s, n_theta)
+
+                    # Perform correlation.
```

Comment:
Would be good to note what the axes are of `corrs` to follow the computations below.

Reply:
Added a note.